src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs

```diff
@@ -102,10 +102,173 @@ IEnumerable<CompletionResult> CompleteArgument(
+                parameterName,
+                wordToComplete,
+                commandAst,
+                fakeBoundParameters);
+
+            return CompleteArgumentWithOptions(ConfigureArgumentCompleterOptions(options));
+        }
+
+        /// <summary>
+        /// Configures argument completer options.
+        /// </summary>
+        /// <param name="options">The options to configure.</param>
+        /// <returns>Configured options.</returns>
+        ArgumentCompleterOptions ConfigureArgumentCompleterOptions(ArgumentCompleterOptions options) => options;
+
+        /// <summary>
+        /// Complete argument with options.
+        /// </summary>
+        /// <param name="options">The options to complete arguments with.</param>
+        /// <returns>A collection of completion results.</returns>
+        IEnumerable<CompletionResult> CompleteArgumentWithOptions(ArgumentCompleterOptions options)
+        {
+            if (!options.ShouldComplete)
+            {
+                yield break;
+            }
+
+            string wordToComplete = options.WordToComplete;
+            string quote = CompletionCompleters.HandleDoubleAndSingleQuote(ref wordToComplete);
+
+            foreach (string value in options.PossibleCompletionValues)
+            {
+                if (value.StartsWith(wordToComplete, StringComparison.OrdinalIgnoreCase))
+                {
+                    string completionText = QuoteCompletionText(value, quote, options.EscapeGlobbingPath);
+                    string toolTip = options.ToolTipMapping?.Invoke(value) ?? value;
+                    string listItemText = options.ListItemTextMapping?.Invoke(value) ?? value;
+
+                    yield return new CompletionResult(completionText, listItemText, options.CompletionResultType, toolTip);
+                }
+            }
+        }
+
+        private static string QuoteCompletionText(string completionText, string quote, bool escapeGlobbingPath)
+        {
+            if (CompletionCompleters.CompletionRequiresQuotes(completionText, escapeGlobbingPath))
+            {
+                string quoteInUse = string.IsNullOrEmpty(quote) ? "'" : quote;
+
+                completionText = quoteInUse == "'"
+                    ? completionText.Replace("'", "''")
+                    : completionText.Replace("`", "``").Replace("$", "`$");
+
+                if (escapeGlobbingPath)
+                {
+                    completionText = quoteInUse == "'"
+                        ? completionText.Replace("[", "`[").Replace("]", "`]")
+                        : completionText.Replace("[", "``[").Replace("]", "``]");
+                }
+
+                return quoteInUse + completionText + quoteInUse;
+            }
+
+            return quote + completionText + quote;
+        }
+
     }
 #nullable restore
 
+    /// <summary>
+    /// Represents options for argument completers.
+    /// </summary>
+    public sealed class ArgumentCompleterOptions
```

Comment:
The `options` pattern assumes there is a code which get the options does _all work_. But we haven't such code. So this is just a decorative pattern and in fact it just imposes how to build custom code. This is wrong. We should not restrict users unless absolutely necessary.
The reason why the interface was originally created rather than the class is precisely to give users maximum freedom in how they build their code.
That's why I recommended that you explore interfaces with defaults, which gives this approach even more power.
Most likely, this will cover all the fundamental needs and we won't even have to think about the auxiliary class.

Reply:
@iSazonov I agree, however if you try and do it all in the interface you run into issues:

- Interfaces cannot have initializers, so default values cannot be set.
- If we define fields in interface, it breaks existing code which relies on the interface since those new fields are expected to be implemented.

Which was the motivation in bringing an auxiliary class in the first place. I don't really see a cleaner way around this.

If our goal is to give users the most freedom but also use sensible default behaviour without writing repetitive code I think a base class is a better option.

Reply:
I guess if we just replace all the fields with methods that have default implementation it could work? Seems messy but that's the only other way I see.

Reply:
You're trying to put absolutely everything in one place. Don't try to embrace the immensity. This is too ambitious a task, comparable to Noah's Ark. 
We need to move from the simple to "the even more useful". I gave an example in the linked issue how enhance the interface. You can use it as start point.

Reply:
@iSazonov Yep I have reconsidered and it seems to work now with some workarounds with static fields. Please review latest changes and let me know if I am on the right track 🙂.